apollo-router/src/plugins/authentication/subgraph.rs

```diff
@@ -192,12 +198,111 @@ pub(crate) struct SigningParams {
 
 #[derive(Clone)]
 pub(crate) struct SigningParamsConfig {
-    credentials_provider: Arc<dyn ProvideCredentials>,
+    credentials_provider: CredentialsProvider,
     region: Region,
     service_name: String,
     subgraph_name: String,
 }
 
+#[derive(Clone, Debug)]
+struct CredentialsProvider {
+    credentials: Arc<RwLock<Credentials>>,
+    _credentials_updater_handle: Arc<JoinHandle<()>>,
+    #[allow(dead_code)]
+    refresh_credentials: Sender<()>,
+}
+
+// Refresh token if it will expire within the next 5 minutes
+const MIN_REMAINING_DURATION: Duration = std::time::Duration::from_secs(60 * 5);
+// If the token doesn't have a validity duration, default to 15 minutes
+const FALLBACK_TOKEN_DURATION: Duration = std::time::Duration::from_secs(60 * 15);
```

Comment:
__Nit__: if a `Credentials` doesn't have an expiry, it means that it never expires. So we could potentially remove the fallback if the `expires_after` value is `None`.

See: https://docs.rs/aws-credential-types/1.2.0/src/aws_credential_types/credentials_impl.rs.html#37

Reply:
I see, and if we manually invalidate the token, subgraph calls will yield 401s so users could still be able to diagnose this. lgtm!